src/components/Table/utils/RowStates.ts

```diff
@@ -67,24 +67,17 @@ export class RowStates {
     this.topRows = rows.filter((row) => row !== headerRow).map((row) => addRowAndChildren(row));
     // And attach them to the header for select-all/etc. to work
     if (this.header) {
-      this.header.children = this.topRows.filter((rs) => !reservedRowKinds.includes(rs.row.kind));
+      this.header.children = [
+        // Always add the keptGroupRow, and we'll use keptGroupRow.isMatched=true/false to keep it
+        // from missing up "header is all selected" if its hidden/when there are no kept rows.
+        this.keptGroupRow,
+        ...this.topRows.filter((rs) => !reservedRowKinds.includes(rs.row.kind)),
+      ];
     }
 
     // Then mark any remaining as removed
     for (const state of existing) state.markRemoved();
 
-    const keptRows = this.keptRows;
-    if (keptRows.length > 0) {
-      // Stitch the current keptRows into the placeholder keptGroupRow
-      this.keptGroupRow.children = keptRows;
-      this.keptGroupRow.row.children = keptRows.map((rs) => rs.row);
```

Comment:
This is the imperative "keep the kept rows up to date" that was, by happenstance, running all the time, so when I fixed "don't run all the time", we had to get this updated in a different way.